### ISSUE TITLE
`torch.embedding` returns random data with indices on meta device

### ISSUE THREAD

**Issue by albertz** [pytorch/pytorch#109781]
### 🐛 Describe the bug

```python
import torch

weight = torch.rand(2,3)  # CPU
indices = torch.zeros(4,5, device="meta", dtype=torch.int32)
y = torch.embedding(weight, indices)
```
Now `y` contains random data.

Actually, I wondered also, why does this not throw an exception, such that it only works when both inputs are on the same device?

Or maybe `torch.embedding` actually supports mixing devices, e.g. `indices` on CPU?
(Slightly off-topic question: What is really the difference between `torch.index_select(weight, dim=0, index=indices.flatten())` + `reshape`, `weight[indices]` and `torch.embedding(weight, indices)`? I don't really know when to use what.)

I'm not really sure what output to expect. Maybe also some tensor on the "meta" device?

### Versions

Collecting environment information...
PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.6.3 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.7 | packaged by conda-forge | (default, Sep 29 2021, 19:24:02)  [Clang 11.1.0 ] (64-bit runtime)
Python platform: macOS-12.6.3-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] evotorch==0.1.0
[pip3] lovely-numpy==0.2.8
[pip3] mypy==0.991
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.24.2
[pip3] torch==2.0.1
[pip3] torch-tb-profiler==0.4.1
[pip3] torchaudio==2.0.2
[pip3] torchdata==0.6.1
[pip3] torchvision==0.15.2
[conda] Could not collect

**Comment by ezyang** [comment]
It is something of a historical accident that we allow mixed meta/regular tensors for embedding/indexing, but unfortunately I can't easily fix it now because people started relying on it to conveniently test embedding tables that are too big to fit in memory by representing them as meta tensors, but still passing regular cpu tensors into the net. An output tensor on meta device would be very reasonable, as would also be erroring. Maybe we should still try to fix this...
